plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,354 @@
+            }
+            else {
+                Map<String, Optional<Partition>> partitions = listAllPartitions(metastore, hiveTable);
+                for (Map.Entry<String, Optional<Partition>> partition : partitions.entrySet()) {
+                    Storage storage = partition.getValue().orElseThrow().getStorage();
+                    log.debug("Building data files from %s", storage.getLocation());
+                    HiveStorageFormat partitionStorageFormat = extractHiveStorageFormat(storage.getStorageFormat());
+                    StructLike partitionData = DataFiles.data(partitionSpec, partition.getKey());
+                    dataFilesBuilder.addAll(buildDataFiles(session, partitionStorageFormat, storage.getLocation(), partitionSpec, partitionData, nameMapping));
+                }
+            }
+
+            log.debug("Start new transaction");
+            Transaction transaction = catalog.newCreateTableTransaction(
+                    session,
+                    sourceTableName,
+                    schema,
+                    parsePartitionFields(schema, toPartitionFields(hiveTable)),
+                    location,
+                    properties);
+
+            List<DataFile> dataFiles = dataFilesBuilder.build();
+            log.debug("Append data %d data files", dataFiles.size());
+            Table table = transaction.table();
+            AppendFiles append = table.newAppend();
+            dataFiles.forEach(append::appendFile);
+            append.commit();
+
+            log.debug("Set preparatory table properties in a metastore for migrations");
+            PrincipalPrivileges principalPrivileges = isUsingSystemSecurity ? NO_PRIVILEGES : buildInitialPrivilegeSet(session.getUser());
+            io.trino.plugin.hive.metastore.Table newTable = io.trino.plugin.hive.metastore.Table.builder(hiveTable)
+                    .setParameter(METADATA_LOCATION_PROP, location)
+                    .setParameter(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(ENGLISH))
+                    .build();
+            metastore.replaceTable(schemaName, tableName, newTable, principalPrivileges);
+
+            transaction.commitTransaction();
+            log.debug("Successfully migrated %s table to Iceberg format", sourceTableName);
+        }
+        catch (Exception e) {
+            throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to migrate table", e);
+        }
+    }
+
+    private Map<String, String> icebergTableProperties(String location, Map<String, String> hiveTableProperties, NameMapping nameMapping)
+    {
+        Map<String, String> icebergTableProperties = new HashMap<>();
+
+        // Copy all Hive table properties so that we can undo the migration easily. This is same as Spark implementation.
+        icebergTableProperties.putAll(hiveTableProperties);
+        EXCLUDED_PROPERTIES.forEach(icebergTableProperties::remove);
```

Comment:
This might be easier to read if we do it inline
```java
icebergTableProperties.remove("path");
icebergTableProperties.remove("transient_lastDdlTime");
icebergTableProperties.remove("serialization.format");